common/setups/rasr/nn_system.py

```diff
@@ -0,0 +1,554 @@
+        pronunciation_scales: List[float],
+        lm_scales: List[float],
+        optimize_am_lm_scale: bool,
+        corpus_key: str,
+        feature_flow_key: str,
+        search_parameters: Dict,
+        lattice_to_ctm_kwargs: Dict,
+        parallelize_conversion: bool,
+        rtf: int,
+        mem: int,
+        **kwargs,
+    ):
+        with tk.block(f"{name}_recognition"):
+            recog_func = self.recog_and_optimize if optimize_am_lm_scale else self.recog
+
+            native_lstm_job = returnn.CompileNativeOpJob("NativeLstm2")
+
+            graph_compile_job = returnn.CompileTFGraphJob(
+                self.adapt_returnn_config_for_recog(returnn_config)
+            )
+
+            forward_output_layer = returnn_config.get("forward_output_layer", "output")
+
+            feature_flow = self.feature_flows[corpus_key][feature_flow_key]
+
+            scorer = rasr.PrecomputedHybridFeatureScorer(
+                prior_mixtures=acoustic_mixture_path,
+                priori_scale=prior_scale,
+            )
+
+            for p, l, cp_cur in itertools.product(
+                pronunciation_scales, lm_scales, checkpoints
+            ):
+                tf_flow = self.get_tf_flow(
+                    cp_cur,
+                    graph_compile_job.out_graph,
+                    native_lstm_job.out_op,
+                    forward_output_layer,
+                )
+                flow = self.add_tf_flow_to_base_flow(feature_flow, tf_flow)
+                flow.config.tf_fwd.loader.saved_model_file = cp_cur
+
+                recog_func(
+                    name=f"{name}-{corpus_key}-ps{p:02.2f}-lm{l:02.2f}",
+                    prefix=f"nn_recognition/{name}/",
+                    corpus=corpus_key,
+                    flow=flow,
+                    feature_scorer=scorer,
+                    pronunciation_scale=p,
+                    lm_scale=l,
+                    search_parameters=search_parameters,
+                    lattice_to_ctm_kwargs=lattice_to_ctm_kwargs,
+                    parallelize_conversion=parallelize_conversion,
+                    rtf=rtf,
+                    mem=mem,
+                    **kwargs,
+                )
+
+    # -------------------- Rescoring  --------------------
+
+    def nn_rescoring(self):
+        # TODO calls rescoring setup
+        raise NotImplementedError
+
+    # -------------------- run functions  --------------------
+
+    def run_data_preparation_step(self, step_args):
+        # TODO here be ogg zip generation for training or lattice generation for SDT
+        raise NotImplementedError
+
+    def run_nn_step(self, step_args):
+        for trn_c, cv_c, dvtr_c, name in itertools.product(
+            self.train_corpora,
+            self.cv_corpora,
+            self.devtrain_corpora,
+            step_args.returnn_configs,
+        ):
+            if (trn_c, cv_c) in self.train_cv_pairing or self.train_cv_pairing is None:
+                continue
```

Comment:
do you really mean to skip over all combinations defined in the pairing or is there a `not` missing?

Reply:
I am wondering if iterating over `itertools.product(..)` is the right thing to do, or if in the step args a list of `(trn_c, cv_c, dvtr_c, name)` tuples can be defined that is iterated over.

Reply:
yes. good point. that could be nicer and maybe more user-friendly..